### DISCUSSION TITLE
How many network requests does a sequential transaction take?

### DISCUSSION THREAD

**Discussion by jeff-an** [prisma/prisma#24343] [Q&A] [answered]
### Question

It seems like interactive Prisma transactions perform a network call to the database for each query executed in the transaction. This makes sense since later queries may depend on the return value of previous ones, so the queries cannot be executed lazily.

But for [sequential prisma client operations](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#sequential-prisma-client-operations) it seems like it would be feasible to generate a single statement and thus complete the transaction in a single network roundtrip. Is this the actual behavior?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @jeff-an 👋 

Thank you for raising this question.

Sequential transactions in Prisma are batched. For example:
```js
const [posts, totalPosts] = await prisma.$transaction([
  prisma.post.findMany({ where: { title: { contains: 'prisma' } } }),
  prisma.post.count(),
])
```
Either all of the operations succeed or none of them succeed. This is done in a single network round trip. Your understanding of this behaviour is correct.

**Reply by jeff-an** [reply]
Thanks for the confirmation!

It would great to call this out in the prisma docs as this performance characteristic is currently not documented. It seems like many people have been bitten by calling tx.update() in a for loop, for example, expecting it to be a single network roundtrip.

**Reply by janpio** [reply, marked as answer]
> This is done in a single network round trip. Your understanding of this behaviour is correct.

This is unfortunately not the case. These Client queries are handled internally in one go, but each query is sent to the database by itself. Prisma generally uses prepared statements, and with those combining queries is often not possible or simple.

Only queries that are internally batched up and executed via a dataloader can overcome this limitation, but that is only implemented for `findUnique()` queries right now.

**Reply by jeff-an** [reply]
It seems like this performance characteristic makes batch updates essentially impossible through Prisma then? The best option seems to just be to template the string myself and use a single $executeRaw statement?

**Reply by janpio** [reply]
Possible, we are still lacking some "bulk" APIs there.

What exactly is your use case - updating multiple entries with different `where` with different `data` at the same time? If you have a SQL example, that would be optimal (and then could also be the basis for a feature request).
